### ISSUE TITLE
Fastlane hangs at the end depending on n/w connectivity

### ISSUE THREAD

**Issue by mgrebenets** [fastlane/fastlane#13510]
### New Issue Checklist

- [x] Updated fastlane to the latest version
- [x] I read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I read [docs.fastlane.tools](https://docs.fastlane.tools)
- [x] I searched for [existing GitHub issues](https://github.com/fastlane/fastlane/issues)

### Issue Description

When no network connectivity is available, fastlane hangs at the end of execution and doesn't timeout unless interrupted (with Ctrl+C).
This occurs both for successful execution and when errors occur.

In my setup I have all the `/HTTP[S]_PROXY/i` environment variables configured and pointing to our company proxy. So when the computer is connected to internal network, fastlane successfully finishes execution and works as expected.

However, if I then connect computer to a 4G hotspot but keep the environment variables unchanged, the company proxy is no longer reachable.
That's OK for my lanes though, because they don't need any n/w connectivity and (I thought) fastlane doesn't need any n/w connectivity either.

The lane is as simple as this:

```ruby
  lane :example_lane do
    sh("echo example_lane")
  end
```
However, that doesn't seem to be the case and at the very end fastlane hangs for indefinite amount of time.

```shell
[10:54:11]: fastlane.tools finished successfully 🎉

# (never finishes execution)
```

After interrupting fastlane with Ctrl+C I can see the following stack trace:

```
^C/Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane_core/lib/fastlane_core/analytics/analytics_session.rb:68:in `join': Interrupt
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane_core/lib/fastlane_core/analytics/analytics_session.rb:68:in `map'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane_core/lib/fastlane_core/analytics/analytics_session.rb:68:in `finalize_session'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane_core/lib/fastlane_core/ui/fastlane_runner.rb:133:in `run!'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/commander-fastlane-4.4.6/lib/commander/delegates.rb:15:in `run!'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane/lib/fastlane/commands_generator.rb:333:in `run'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane/lib/fastlane/commands_generator.rb:41:in `start'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/fastlane/lib/fastlane/cli_tools_distributor.rb:108:in `take_off'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/gems/fastlane-2.105.2/bin/fastlane:20:in `<top (required)>'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/bin/fastlane:23:in `load'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/bin/fastlane:23:in `<main>'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:24:in `eval'
        from /Users/bigmac/.rvm/gems/ruby-2.3.3/bin/ruby_executable_hooks:24:in `<main>'
```

I've examined the contents of `analytics_session.rb` and it's this:

```ruby
    def finalize_session
      @threads.map(&:join)
    end
```

Looks like it waits for some particular thread to finish.

I suspect the thread in question has something to do with analytics, which leads to my next question:

> I've used `opt_out_usage` in my Fastfile, why is it still running analytics session?

I've also configured the following in `before_all` hook to make sure the number of external n/w calls is reduced:

```ruby
    ENV["FASTLANE_HIDE_GITHUB_ISSUES"] = "YES"
    ENV["FASTLANE_SKIP_UPDATE_CHECK"] = "YES"
    ENV["FASTLANE_HIDE_CHANGELOG"] = "YES"
```
The problem also happens on our CI agents, which have limited external connectivity.

So to sum it up.

- Should it not try to capture any analytics if I choose to opt out of usage?
- Even if there are some outgoing connections, should be there a timeout setting to avoid hanging execution?

##### Complete output when running fastlane, including the stack trace and command used

The irony is that I can't capture output, since `bundle exec fastlane example_lane --capture_output` never finishes :)

### Environment

I was able to reproduce this both with Ruby 2.3.3 and 2.4.1.
The stack is for 2.4.1.

<details>
  <pre><details><summary>✅ fastlane environment ✅</summary>

### Stack

| Key                         | Value                                             |
| --------------------------- | ------------------------------------------------- |
| OS                          | 10.14.1                                           |
| Ruby                        | 2.4.1                                             |
| Bundler?                    | true                                              |
| Git                         | git version 2.15.2 (Apple Git-101.1)              |
| Installation Source         | ~/.rvm/gems/ruby-2.4.1/bin/fastlane               |
| Host                        | Mac OS X 10.14.1 (18B45d)                         |
| Ruby Lib Dir                | ~/.rvm/rubies/ruby-2.4.1/lib                      |
| OpenSSL Version             | OpenSSL 1.1.0g  2 Nov 2017                        |
| Is contained                | false                                             |
| Is homebrew                 | false                                             |
| Is installed via Fabric.app | false                                             |
| Xcode Path                  | /Applications/Xcode-9.4.1.app/Contents/Developer/ |
| Xcode Version               | 9.4.1                                             |

### System Locale

| Variable | Value       |   |
| -------- | ----------- | - |
| LANG     | en_GB.UTF-8 | ✅ |
| LC_ALL   |             |   |
| LANGUAGE |             |   |

### fastlane files:

<details><summary>`./fastlane/Fastfile`</summary>

```ruby
# Sample

opt_out_usage

default_platform(:ios)

platform :ios do
  before_all do
    ENV["FASTLANE_HIDE_GITHUB_ISSUES"] = "YES" # Disable GitHub issue lookup
    ENV["FASTLANE_SKIP_UPDATE_CHECK"] = "YES"
    ENV["FASTLANE_HIDE_CHANGELOG"] = "YES"
  end

  desc "Description of what the lane does"
  lane :example_lane do
    sh("echo example_lane")
  end
end

```
</details>

<details><summary>`./fastlane/Appfile`</summary>

```ruby
# app_identifier("[[APP_IDENTIFIER]]") # The bundle identifier of your app
# apple_id("[[APPLE_ID]]") # Your Apple email address


# For more information about the Appfile, see:
#     https://docs.fastlane.tools/advanced/#appfile

```
</details>

### fastlane gems

| Gem      | Version | Update-Status  |
| -------- | ------- | -------------- |
| fastlane | 2.105.2 | 💥 Check failed |

### Loaded fastlane plugins:

**No plugins Loaded**

<details><summary><b>Loaded gems</b></summary>

| Gem                       | Version      |
| ------------------------- | ------------ |
| did_you_mean              | 1.1.0        |
| io-console                | 0.4.6        |
| CFPropertyList            | 3.0.0        |
| public_suffix             | 2.0.5        |
| addressable               | 2.5.2        |
| atomos                    | 0.1.3        |
| babosa                    | 1.0.2        |
| bundler                   | 1.16.2       |
| claide                    | 1.0.2        |
| colored                   | 1.2          |
| colored2                  | 3.1.2        |
| highline                  | 1.7.10       |
| commander-fastlane        | 4.4.6        |
| declarative               | 0.0.10       |
| declarative-option        | 0.1.0        |
| unf_ext                   | 0.0.7.5      |
| unf                       | 0.1.4        |
| domain_name               | 0.5.20180417 |
| dotenv                    | 2.5.0        |
| emoji_regex               | 0.1.1        |
| excon                     | 0.62.0       |
| multipart-post            | 2.0.0        |
| faraday                   | 0.15.3       |
| http-cookie               | 1.0.3        |
| faraday-cookie_jar        | 0.0.6        |
| faraday_middleware        | 0.12.2       |
| fastimage                 | 2.1.4        |
| gh_inspector              | 1.1.3        |
| jwt                       | 2.1.0        |
| memoist                   | 0.16.0       |
| multi_json                | 1.13.1       |
| os                        | 1.0.0        |
| signet                    | 0.10.0       |
| googleauth                | 0.6.6        |
| httpclient                | 2.8.3        |
| mime-types-data           | 3.2018.0812  |
| mime-types                | 3.2.2        |
| uber                      | 0.1.0        |
| representable             | 3.0.4        |
| retriable                 | 3.1.2        |
| google-api-client         | 0.23.9       |
| json                      | 2.1.0        |
| mini_magick               | 4.5.1        |
| multi_xml                 | 0.6.0        |
| plist                     | 3.4.0        |
| rubyzip                   | 1.2.2        |
| security                  | 0.1.3        |
| naturally                 | 2.2.0        |
| simctl                    | 1.6.5        |
| slack-notifier            | 2.3.2        |
| terminal-notifier         | 1.8.0        |
| unicode-display_width     | 1.4.0        |
| terminal-table            | 1.8.0        |
| tty-screen                | 0.6.5        |
| tty-cursor                | 0.6.0        |
| tty-spinner               | 0.8.0        |
| word_wrap                 | 1.0.0        |
| nanaimo                   | 0.2.6        |
| xcodeproj                 | 1.6.0        |
| rouge                     | 2.0.7        |
| xcpretty                  | 0.3.0        |
| xcpretty-travis-formatter | 1.0.0        |
</details>

*generated on:* **2018-10-10**
</details></pre>
</details>

**Comment by mgrebenets** [comment]
Some updates.

It may be related to Xcode version as well somehow.
One of our CI jobs is using Fastlane `2.99.1`.
The same lanes finish when running builds using Xcode 9.4.1 (although it takes almost 4 minutes before Fastlane wraps up:

```
[12:14:39]	[Step 4/6] [12:14:40]: fastlane.tools just saved you 19 minutes!   
[12:18:42]	[Step 4/6] Process exited with code 0
```

While jobs using Xcode 10 simply hang at this point and never terminate.

**Comment by mgrebenets** [comment]
Setting up env vars for proxy in `before_all` didn't help:

```ruby
    ENV["HTTP_PROXY"] = nil
    ENV["HTTPS_PROXY"] = nil
    ENV["http_proxy"] = nil
    ENV["https_proxy"] = nil
```

**Comment by mgrebenets** [comment]
OK, so digging a bit deeper, I've found that in `AnalyticsIngesterClient` the post even will not be created if the following condition is met:

```ruby
      if Helper.test? || FastlaneCore::Env.truthy?("FASTLANE_OPT_OUT_USAGE")
        return nil
      end
```

the `opt_out_usage` _does_ set `FASTLANE_OPT_OUT_USAGE` to `YES`, but it does it **after** the analytics session has been initiallized, which is why the whole problem happens.
I've added debug print messages to `post_event` in `analytics_ingester_client.rb` and to `opt_out_usage.rb` and I can see from the logs that ingester is initialized before call to `opt_out_usage`.

P.S.
As a work-around, prefixing fastlane commands with `FASTLANE_OPT_OUT_USAGE=YES` solves the problem.
E.g. in our case we've set it on CI to apply to all jobs.
